internal/allocator/allocator.go

```diff
@@ -305,44 +305,29 @@ func (a *Allocator) Allocate(svcKey string, svc *v1.Service, serviceIPFamily ipf
 // This method returns sorted ip pools which are allocatable for given service.
 // When ip pool is not set with priority, then just append it after sorted
 // priority ip pools.
-func (a *Allocator) sortedPoolsForService(svc *v1.Service) []*config.Pool {
-	var priorityPools, noPriorityPools []*config.Pool
+func (a *Allocator) pinnedPoolsForService(svc *v1.Service) []*config.Pool {
+	var pools []*config.Pool
 	if svc == nil {
-		return priorityPools
+		return pools
 	}
 	for _, nsPoolName := range a.pools.ByNamespace[svc.Namespace] {
```

Comment:
why isn't it enough to just iterate over all of the pools ByName instead of multiple fors? the `isPoolCompatible` would filter out the unrelated pools anyways? (sorry if I'm confusing stuff, not familiar with this part and the Pools struct)

Reply:
yep, but here we save some loops by skipping the non pinned loops, as "byName" contains all of them

Reply:
not sure about saving loops, as it seems the current implementation can add the same pool twice to the `pools` slice. I think it would be more readable to just iterate once over everything and the performance hit (if at all) shouldn't be too big

Reply:
we can avoid pool duplicates in the `pools` slice with a change in config.go, but that was reverted based on this [comment](https://github.com/metallb/metallb/pull/1821#discussion_r1107011002). I think iterating over all pools may be somewhat costlier than current one as it needs to validate ns and svc labels for every ns/svc pools.

Reply:
oh! no, the isPoolCompatibleWithService ensures no duplicates get added into `pools` slice, i think we're better here.

Reply:
what I meant (and I might be wrong), that in the current way byNS/bySelector might overlap (or contain everything?) and that way we aren't really saving any iterations by not going once over everything

Reply:
byNS/bySelector may contain subset or all of ip pools. but it depends on user's pool config. If we do it by iterating over all pools via `a.pools.ByName`, the lookup cost for desired ns/svc pools are high. @fedepaol your thoughts ?

Reply:
iiuc, if most of the people won't use this feature the byNS/bySelector will always contain all of the pools.
having said that, I wouldn't want this to block this PR if we intend to have it in the next release as it's not too crucial.

Reply:
I can't think about a scenario where there are more than... 100 pools, to say a number.
On the other hand, I think the current version is more than understandable so I'd say let's go with this version, I don't mind cleaning up later if it becomes too messy / if we add another discriminator.